src/components/posts/view-post/PostPage.tsx

```diff
@@ -37,27 +40,27 @@ export const PostPage: NextPage<PostDetailsProps> = ({ postDetails: initialPost,
 
   if (!space || isHiddenSpace(space.struct)) return <PostNotFound />
 
-  const { struct: initStruct, content } = post;
+  const { struct: initStruct, content } = post
 
-  if (!content) return null;
+  if (!content) return null
 
   const { isNotMobile } = useResponsiveSize()
   const struct = useSubscribedPost(initStruct)
   const postDetails = { ...initialPost, post: { struct, content } }
 
   const spaceData = space || postDetails.space || useLoadUnlistedSpace(struct.owner).myHiddenSpace
 
-  const { title, body, image, canonical, tags } = content;
-
   if (!spaceData) return <Loading />
 
-  const spaceStruct = spaceData.struct;
+  const spaceStruct = spaceData.struct
+
+  const { title, body, image, canonical = fullPath(postUrl(spaceStruct, postDetails.post)), tags } = content
```

Comment:
Create a canonical URL before this line, because now this line is ugly huge.
```suggestion
  const canonical = content.canonical || fullPath(postUrl(spaceStruct, postDetails.post))
  const { title, body, image, tags } = content
```